File: javascript/178-merge_sort.js
var Mocha = require('mocha');
var assert = require('chai').assert;
var expect = require('chai').expect;
var mocha = new Mocha();


function mergeSort(input_array) {
    if (!(input_array instanceof Array)) {
        throw TypeError("An array is required for mergeSort")
    }
    
    if (input_array.length === 1) {
        return input_array
    }
    
    let midPoint = Math.ceil(input_array.length / 2)
    let left = input_array.slice(0, midPoint)
    let right = input_array.slice(midPoint)

    return mergeArray(
        mergeSort(left), 
        mergeSort(right)
    )

}

function mergeArray(array1, array2) {

    if (!(array1 instanceof Array) || ((array2 !== undefined) && !(array2 instanceof Array))) {
        throw TypeError("An array is required for mergeArray")
    }
    if (!array2) {
        return array1
    }
        let output_array = []
        let i = 0
        let j = 0
        while ((i < array1.length) || (j < array2.length)) {
            let left_first = array1[i]
            let right_first = array2[j]
            if (right_first < left_first || left_first === undefined) {
                j++;
                output_array.push(right_first);
            }
            else {
                i++;
                output_array.push(left_first);
            }
        }
        return output_array
    // }
}

describe('mergeArray', () => {
    context('Input checks', () => {
        it('should throw a TypeError if no value passed', () => {
            expect(() => {mergeArray()}).to.throw(TypeError, "An array is required for mergeArray");
        })

        it('should throw a TypeError if value1 is not Array instance', () => {
            expect(() => {mergeArray(1, [1, 2, 3])}).to.throw(TypeError, "An array is required for mergeArray");
        })

        it('should throw a TypeError if value2 is not Array instance', () => {
            expect(() => {mergeArray([1, 2, 3], 1)}).to.throw(TypeError, "An array is required for mergeArray");
        })

        it('should not throw a TypeError if both parameters Array instances', () => {
            expect(() => {mergeArray([1,2,3], [1,2,3])}).to.not.throw(TypeError, "An array is required for mergeArray");
        })

        it('should not throw a TypeError if only parameter is passed', () => {
            expect(() => {mergeArray([1,2,3])}).to.not.throw(TypeError, "An array is required for mergeArray");
        })
    })

    context('merge Arrays', () => {
        it('should return the input if one array passed', () => {
            let testArray = [1]
            let expectedArray = JSON.stringify(testArray)
            let actualArray = JSON.stringify(mergeArray(testArray))
            expect(actualArray).to.equal(expectedArray)
        })

        it('should merge two arrays of length === 1', () => {
            let expectedArray = JSON.stringify([1, 2])
            let actualArray = JSON.stringify(mergeArray([1], [2]))
            expect(actualArray).to.equal(expectedArray)
        })

        it('should merge arrays of length === 2 & 1', () => {
            let expectedArray = JSON.stringify([1, 2, 3])
            let actualArray = JSON.stringify(mergeArray([1, 2], [3]))
            expect(actualArray).to.equal(expectedArray)
        })

        it('should merge two arrays of length === 2', () => {
            let expectedArray = JSON.stringify([1, 2, 3, 4])
            let actualArray = JSON.stringify(mergeArray([1, 2], [3, 4]))
            expect(actualArray).to.equal(expectedArray)
        })

        it('should merge and sort two arrays of length === 2', () => {
            let expectedArray = JSON.stringify([1, 2, 3, 4])
            let actualArray = JSON.stringify(mergeArray([2, 4], [1, 3]))
            expect(actualArray).to.equal(expectedArray)
        })
    })
})


describe('MergeSort ', () => {
    context('Input checks', () => {
        it('should throw a TypeError if no value passed', () => {
            expect(() => {mergeSort()}).to.throw(TypeError, "An array is required for mergeSort");
        })

        it('should throw TypeError if non array passed', () => {
            expect(() => {mergeSort(Object())}).to.throw(TypeError, "An array is required for mergeSort");

        })
    })


    context('Number sorting', () => {
        it('should sort a list of length 1', () => {
            let actual = JSON.stringify(mergeSort([1]))
            let expected_output = JSON.stringify([1])
            expect(actual).to.equal(expected_output)
        })

        it('should not change the ordering of a pre-sorted list of length 3', () => {
            let actual = JSON.stringify(mergeSort([1, 2, 3]))
            let expected_output = JSON.stringify([1, 2, 3])
            expect(actual).to.equal(expected_output)
        })

        it('should re-order a reverse sorted list of length 3', () => {
            let actual = JSON.stringify(mergeSort([3, 2, 1]))
            let expected_output = JSON.stringify([1, 2, 3])
            expect(actual).to.equal(expected_output)
        })

        it('should re-order a reverse sorted list of length 6', () => {
            let actual = JSON.stringify(mergeSort([6, 5, 4, 3, 2, 1]))
            let expected_output = JSON.stringify([1, 2, 3, 4, 5, 6])
            expect(actual).to.equal(expected_output)
        })

        it('should re-order a randomly sorted list of length 8', () => {
            let actual = JSON.stringify(mergeSort([6, 5, 3, 1, 8, 7, 2, 4]))
            let expected_output = JSON.stringify([1, 2, 3, 4, 5, 6, 7, 8])
            expect(actual).to.equal(expected_output)
        })

        it('should re-order a randomly sorted list of length 11', () => {
            let actual = JSON.stringify(mergeSort([99, 44, 6, 2, 1, 5, 63, 87, 283, 4, 0]))
            let expected_output = JSON.stringify([0, 1, 2, 4, 5, 6, 44, 63, 87, 99, 283])
            expect(actual).to.equal(expected_output)
        })
    })
})


mocha.run()

File: python/178-merge_sort.py
import math
import unittest


def merge_sort_recursive(input_array):
    if len(input_array) == 1:
        return input_array

    mid_index = math.ceil(len(input_array) / 2)
    left = input_array[:mid_index]
    right = input_array[mid_index:]

    return merge_array(
        merge_sort_recursive(left),
        merge_sort_recursive(right)
    )


def merge_array(array1, array2=None):
    if not array2:
        return array1
    i, j = 0, 0
    output_array = []
    while i < len(array1) or j < len(array2):
        if j == len(array2):
            output_array.append(array1[i])
            i += 1
        elif i == len(array1):
            output_array.append(array2[j])
            j += 1
        elif array1[i] < array2[j]:
            output_array.append(array1[i])
            i += 1
        else:
            output_array.append(array2[j])
            j += 1
    return output_array


class TestMergeArray(unittest.TestCase):

    @classmethod
    def setUpClass(cls) -> None:
        cls.test_cases = [
            ([[1]], [1]),
            ([[1, 2], [3]], [1, 2, 3]),
            ([[1, 2], [3, 4]], [1, 2, 3, 4]),
            ([[2, 4], [1, 3]], [1, 2, 3, 4]),
        ]

    def test_no_parameter_raises(self):
        with self.assertRaises(TypeError):
            merge_array()

    def test_merge_array(self):
        for test_case, expected_result in self.test_cases:
            with self.subTest(f'test case: {test_case}'):
                actual = merge_array(*test_case)
                self.assertEqual(actual, expected_result)

class TestMergeSort(unittest.TestCase):

    @classmethod
    def setUpClass(cls) -> None:
        cls.test_cases = [
            ([1], [1]),
            ([1, 2, 3], [1, 2, 3]),
            ([3, 2, 1], [1, 2, 3]),
            ([6, 5, 4, 3, 2, 1], [1, 2, 3, 4, 5, 6]),
            ([6, 5, 3, 1, 8, 7, 2, 4], [1, 2, 3, 4, 5, 6, 7, 8]),
            ([1, 2, 3, 5, 6, 7, 8, 4], [1, 2, 3, 4, 5, 6, 7, 8]),
            (
                [99, 44, 6, 2, 1, 5, 63, 87, 283, 4, 0],
                [0, 1, 2, 4, 5, 6, 44, 63, 87, 99, 283]
             )
        ]

        cls.functions_to_test = [merge_sort_recursive, ] #merge_sort_iterative]

    def test_no_parameter_raises(self):
        for function_to_test in self.functions_to_test:
            with self.subTest(f"Function: {function_to_test.__name__}"):
                with self.assertRaises(TypeError):
                    function_to_test()

    def test_merge_sort(self):
        for function_to_test in self.functions_to_test:
            for test_case, expected_result in self.test_cases:
                with self.subTest(f"function: {function_to_test.__name__}; test case: {test_case}"):
                    actual = function_to_test(test_case)
                    self.assertEqual(actual, expected_result)


if __name__ == "__main__":
    unittest.main()